File: wire/bench/bench.ts
// Wire benchmark in JavaScript with node:net (see README.md).
import { createConnection, createServer, type Socket } from "node:net";

const PORT = 38501;
const N = 1_048_576;
const R = 64;
const CHK = 3187671040;

function fill(b: Uint8Array): void {
  for (let i = 0; i < N; i++) b[i] = (i * 31 + 7) & 255;
}

function hashBuf(h: number, b: Uint8Array): number {
  for (let i = 0; i < b.length; i++) h = (Math.imul(h, 31) + b[i]) >>> 0;
  return h >>> 0;
}

function readExact(sock: Socket, n: number): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const out = Buffer.alloc(n);
    let off = 0;
    const onData = (c: Buffer) => {
      c.copy(out, off);
      off += c.length;
      if (off >= n) {
        sock.off("data", onData);
        sock.off("error", reject);
        resolve(out);
      }
    };
    sock.on("data", onData);
    sock.on("error", reject);
  });
}

function writeAll(sock: Socket, buf: Buffer): Promise<void> {
  return new Promise((resolve, reject) => {
    sock.write(buf, (err) => (err ? reject(err) : resolve()));
  });
}

async function main(): Promise<void> {
  const ready = Promise.withResolvers<void>();
  const srv = createServer((conn) => {
    (async () => {
      for (let r = 0; r < R; r++) {
        const data = await readExact(conn, N);
        await writeAll(conn, data);
      }
      conn.end();
    })().catch(() => conn.destroy());
  });
  srv.on("error", (e) => ready.reject(e));
  srv.listen(PORT, "127.0.0.1", () => ready.resolve());
  await ready.promise;

  const raw = new Uint8Array(N);
  fill(raw);
  const buf = Buffer.from(raw.buffer, raw.byteOffset, raw.byteLength);

  await new Promise<void>((resolve, reject) => {
    const conn = createConnection({ host: "127.0.0.1", port: PORT }, async () => {
      try {
        let cs = 0;
        const t0 = performance.now();
        for (let r = 0; r < R; r++) {
          await writeAll(conn, buf);
          const got = await readExact(conn, N);
          cs = hashBuf(cs, new Uint8Array(got.buffer, got.byteOffset, got.byteLength));
        }
        const ms = performance.now() - t0;
        console.log(`echo\t${ms.toFixed(3)}\t${cs}`);
        conn.end();
        srv.close();
        if (cs !== CHK) reject(new Error("checksum"));
        else resolve();
      } catch (e) {
        reject(e);
      }
    });
    conn.on("error", reject);
  });
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: wire/bench/bench.rs
// Wire benchmark in Rust with std::net (see README.md).
use std::io::{Read, Write};
use std::net::{SocketAddr, TcpListener, TcpStream};
use std::sync::mpsc;
use std::thread;
use std::time::Instant;

const PORT: u16 = 38501;
const N: usize = 1_048_576;
const R: usize = 64;
const CHK: u32 = 3187671040;

fn fill(b: &mut [u8]) {
	for (i, byte) in b.iter_mut().enumerate() {
		*byte = ((i as u64 * 31 + 7) & 255) as u8;
	}
}

fn hash_buf(mut h: u32, b: &[u8]) -> u32 {
	for &byte in b {
		h = h.wrapping_mul(31).wrapping_add(u32::from(byte));
	}
	h
}

fn addr() -> SocketAddr {
	format!("127.0.0.1:{}", PORT).parse().unwrap()
}

fn main() {
	let (ready_tx, ready_rx) = mpsc::channel();
	let th = thread::spawn(move || {
		let ls = TcpListener::bind(addr()).unwrap();
		ready_tx.send(()).unwrap();
		let (mut conn, _) = ls.accept().unwrap();
		let mut buf = vec![0u8; N];
		for _ in 0..R {
			conn.read_exact(&mut buf).unwrap();
			conn.write_all(&buf).unwrap();
		}
	});
	ready_rx.recv().unwrap();

	let mut buf = vec![0u8; N];
	fill(&mut buf);
	let mut conn = TcpStream::connect(addr()).unwrap();
	let mut cs = 0u32;

	let t0 = Instant::now();
	for _ in 0..R {
		conn.write_all(&buf).unwrap();
		conn.read_exact(&mut buf).unwrap();
		cs = hash_buf(cs, &buf);
	}
	let ms = t0.elapsed().as_secs_f64() * 1000.0;
	println!("echo\t{}\t{}", ms, cs);

	th.join().unwrap();
	if cs != CHK {
		std::process::exit(2);
	}
}
